File: apps/ios/Sources/QuranRecitationApp/Features/Practice/PracticeViewModel.swift
import Foundation
import SwiftUI

/// State container for the Practice screen. PR 12 lays the foundation
/// (idle scaffold, ayah loading, dependency wiring); PR 13 adds teacher
/// reference playback, PR 14 records, and PR 15 wires the
/// stop → upload → analyse → result transition end-to-end.
@MainActor
final class PracticeViewModel: ObservableObject {
  @Published private(set) var state: PracticeRecordingState = .idle
  @Published private(set) var ayah: AyahDetail?
  @Published private(set) var surah: SurahSummary?
  @Published private(set) var teacherState: TeacherReferencePanel.State = .idle

  /// Whether the teacher reference repeats on finish. Mirrors the
  /// web `TeacherPanel` loop control. Driven by the loop toggle in
  /// `TeacherReferencePanel`.
  @Published private(set) var isLoopEnabled: Bool = false

  /// Reciter name surfaced on the teacher panel. Hard-coded today —
  /// when the BFF starts returning reciter metadata, this becomes a
  /// `@Published` populated by `loadReference()`.
  let reciterName: String = "Husary Mu'allim"

  let surahId: String
  let initialAyahNumber: Int

  private let backend: QuranBackend
  private let referenceClient: ReferenceAudioClient
  let recorder: AudioRecorder
  let player: AudioPlayer
  private let historyStore: HistoryStore
  private let telemetry: Telemetry
  private var playbackRate: Float = 1.0

  init(
    surahId: String,
    ayahNumber: Int,
    backend: QuranBackend,
    referenceClient: ReferenceAudioClient,
    recorder: AudioRecorder,
    player: AudioPlayer,
    historyStore: HistoryStore,
    telemetry: Telemetry
  ) {
    self.surahId = surahId
    self.initialAyahNumber = ayahNumber
    self.currentAyahNumber = ayahNumber
    self.backend = backend
    self.referenceClient = referenceClient
    self.recorder = recorder
    self.player = player
    self.historyStore = historyStore
    self.telemetry = telemetry
  }

  /// Load the surah header + the active ayah. Called by `PracticeView`
  /// in `.task`. Failures populate `state = .error(...)` so the View
  /// renders a recoverable empty surface.
  func load() async {
    do {
      async let surah = backend.surah(id: surahId)
      async let ayah = backend.ayah(surahId: surahId, ayahNumber: currentAyahNumber)
      self.surah = try await surah
      self.ayah = try await ayah
      if self.ayah == nil {
        let error = AppError.backendUnavailable(operation: "ayah")
        telemetry.error(error, context: ["screen": "practice"])
        state = .error(error)
      }
    } catch let error as AppError {
      telemetry.error(error, context: ["screen": "practice"])
      state = .error(error)
    } catch {
      let appError = AppError.network(underlying: error)
      telemetry.error(appError, context: ["screen": "practice"])
      state = .error(appError)
    }
  }

  /// Currently shown ayah number. Mutates via `goToPreviousAyah()` /
  /// `goToNextAyah()`. Reset to `initialAyahNumber` on construction
  /// so deep-link entries (e.g. Library Continue) land where the
  /// caller expects.
  @Published private(set) var currentAyahNumber: Int

  /// True when there is a previous ayah to navigate to. The UI uses
  /// this to disable the prev button on ayah 1.
  var canGoToPreviousAyah: Bool { currentAyahNumber > 1 }

  /// True when there is a next ayah to navigate to. The UI uses
  /// this to disable the next button on the last ayah of the surah.
  /// Reads `surah?.ayahCount` — until the header has loaded, the
  /// next button is conservatively disabled.
  var canGoToNextAyah: Bool {
    guard let count = surah?.ayahCount else { return false }
    return currentAyahNumber < count
  }

  /// Step back one ayah and reload its text and reference audio.
  /// No-op if already on ayah 1. The teacher player is stopped first
  /// so a half-played previous-ayah reference can't bleed into the
  /// new one.
  func goToPreviousAyah() async {
    guard canGoToPreviousAyah else { return }
    await moveTo(ayahNumber: currentAyahNumber - 1, direction: "prev")
  }

  /// Step forward one ayah and reload its text and reference audio.
  /// No-op once past the surah's final ayah.
  func goToNextAyah() async {
    guard canGoToNextAyah else { return }
    await moveTo(ayahNumber: currentAyahNumber + 1, direction: "next")
  }

  private func moveTo(ayahNumber: Int, direction: String) async {
    player.stop()
    if case let .ready(_, _, _, rate) = teacherState {
      // Step out of `.ready` so the panel shows the loading state
      // until the new reference resolves.
      teacherState = .loading
      _ = rate
    } else {
      teacherState = .loading
    }
    currentAyahNumber = ayahNumber
    telemetry.event(
      TelemetryEvent.practiceAyahNavigated,
      attributes: [
        "surah_id": surahId,
        "ayah": String(ayahNumber),
        "direction": direction
      ]
    )
    do {
      let nextAyah = try await backend.ayah(surahId: surahId, ayahNumber: ayahNumber)
      self.ayah = nextAyah
    } catch let error as AppError {
      telemetry.error(error, context: ["screen": "practice", "scope": "ayah-nav"])
      state = .error(error)
    } catch {
      let appError = AppError.network(underlying: error)
      telemetry.error(appError, context: ["screen": "practice", "scope": "ayah-nav"])
      state = .error(appError)
    }
    await loadReference()
  }

  // MARK: - Teacher reference (PR 13)

  /// Fetch the teacher reference URL and load it into `AudioPlayer`.
  /// Failures route to the panel's `.unavailable` state and log to
  /// telemetry — the user can still record their attempt.
  func loadReference() async {
    teacherState = .loading
    do {
      let reference = try await referenceClient.referenceAudio(
        surahId: surahId,
        ayahNumber: currentAyahNumber
      )
      try player.load(url: reference.url)
      // load() builds a fresh AVAudioPlayer so the previous loop
      // flag is lost; re-apply it here so the loop toggle survives
      // ayah changes (and future reloads).
      player.setLoopEnabled(isLoopEnabled)
      teacherState = .ready(
        duration: player.duration,
        isPlaying: false,
        currentTime: 0,
        rate: playbackRate
      )
    } catch let error as AppError {
      telemetry.error(error, context: ["screen": "practice", "scope": "reference"])
      teacherState = .unavailable
    } catch {
      let appError = AppError.network(underlying: error)
      telemetry.error(appError, context: ["screen": "practice", "scope": "reference"])
      teacherState = .unavailable
    }
  }

  func toggleReferencePlayback() {
    guard case let .ready(duration, isPlaying, currentTime, rate) = teacherState else { return }
    if isPlaying {
      player.pause()
    } else {
      player.play(rate: rate)
      telemetry.event(
        TelemetryEvent.practiceReferencePlayed,
        attributes: ["surah_id": surahId, "ayah": String(currentAyahNumber)]
      )
    }
    teacherState = .ready(
      duration: duration,
      isPlaying: !isPlaying,
      currentTime: currentTime,
      rate: rate
    )
  }

  /// Set the teacher-audio playback rate explicitly. Mirrors the web
  /// `TeacherPanel` 0.75× / 1× / 1.25× pill row. Values outside the
  /// supported set are clamped to the nearest neighbour so a future
  /// caller (e.g. accessibility settings) can't drift the rate into
  /// territory the player doesn't handle gracefully.
  func setReferenceRate(_ rate: Float) {
    let clamped = Self.snapToSupportedRate(rate)
    if clamped == playbackRate { return }
    playbackRate = clamped
    if case let .ready(duration, isPlaying, currentTime, _) = teacherState {
      teacherState = .ready(
        duration: duration,
        isPlaying: isPlaying,
        currentTime: currentTime,
        rate: clamped
      )
      if isPlaying {
        player.play(rate: clamped)
      }
    }
    telemetry.event(
      TelemetryEvent.practiceRateChanged,
      attributes: ["rate": String(format: "%.2f", clamped)]
    )
  }

  /// Supported teacher-audio playback rates. Centralised so the pill
  /// row, telemetry, and the snap-to-nearest helper all stay in sync.
  static let supportedReferenceRates: [Float] = [0.75, 1.0, 1.25]

  /// Toggle the loop-ayah behaviour. When on, the teacher reference
  /// repeats forever on finish; when off it stops and resets. The
  /// new state is pushed straight into `AudioPlayer.numberOfLoops`
  /// — the next reference load reuses the same flag because the
  /// player is shared.
  func toggleLoop() {
    isLoopEnabled.toggle()
    player.setLoopEnabled(isLoopEnabled)
    telemetry.event(
      TelemetryEvent.practiceLoopToggled,
      attributes: ["enabled": String(isLoopEnabled)]
    )
  }

  private static func snapToSupportedRate(_ rate: Float) -> Float {
    supportedReferenceRates.min(by: { abs($0 - rate) < abs($1 - rate) }) ?? 1.0
  }

  // MARK: - Recording (PR 14)

  /// Tap handler for the circular record button. Toggles between idle
  /// and recording. The actual stop → upload → score flow lands in PR 15.
  func toggleRecording() {
    switch state {
    case .idle, .error, .done:
      startRecording()
    case .recording:
      stopRecording()
    case .uploading, .analysing:
      // Ignore taps while a job is in flight; the View should also
      // disable the button via PracticeRecordingState.isBusy.
      return
    }
  }

  private func startRecording() {
    do {
      try recorder.startRecording()
      state = .recording(meters: [], duration: 0)
      observeRecorder()
      telemetry.event(
        TelemetryEvent.practiceRecordingStarted,
        attributes: ["surah_id": surahId, "ayah": String(currentAyahNumber)]
      )
    } catch let error as AppError {
      telemetry.error(error, context: ["screen": "practice"])
      state = .error(error)
    } catch {
      let appError = AppError.audioRecordingFailed(underlying: error)
      telemetry.error(appError, context: ["screen": "practice"])
      state = .error(appError)
    }
  }

  private func stopRecording() {
    let durationMs = Int(recorder.duration * 1000)
    do {
      let recordingURL = try recorder.stopRecording()
      telemetry.event(
        TelemetryEvent.practiceRecordingStopped,
        attributes: ["duration_ms": String(durationMs)]
      )
      Task { await processRecording(at: recordingURL, durationMs: durationMs) }
    } catch let error as AppError {
      telemetry.error(error, context: ["screen": "practice"])
      state = .error(error)
    } catch {
      let appError = AppError.audioRecordingFailed(underlying: error)
      telemetry.error(appError, context: ["screen": "practice"])
      state = .error(appError)
    }
  }

  /// Stop → upload → create job → poll. State transitions to `.uploading`,
  /// then `.analysing(step:)`, then `.done` (or `.error`). On success the
  /// final result is appended to `route` for the View to consume, and
  /// the attempt is persisted via `HistoryStore`.
  @Published var route: [PracticeRoute] = []

  /// Test-only entry point so unit tests can exercise the upload →
  /// score → telemetry pipeline without holding a real recording.
  /// Production code goes through `toggleRecording()`.
  func testProcessRecording(at fileURL: URL, durationMs: Int) async {
    await processRecording(at: fileURL, durationMs: durationMs)
  }

  private func processRecording(at fileURL: URL, durationMs: Int) async {
    state = .uploading
    do {
      let signedUpload = try await telemetry.measure(TelemetryEvent.practiceUploadCompleted) {
        try await backend.requestSignedUploadUrl(
          filename: fileURL.lastPathComponent,
          contentType: "audio/m4a"
        )
      }
      try await backend.uploadAudio(fileURL: fileURL, to: signedUpload.url)

      state = .analysing(step: .transcribing)
      let job = try await backend.createScoringJob(
        uploadKey: signedUpload.uploadKey,
        surahId: surahId,
        ayahNumber: currentAyahNumber
      )
      state = .analysing(step: .comparing)

      let result = try await telemetry.measure(TelemetryEvent.practiceScoringCompleted) {
        try await backend.pollScoringResult(jobId: job.jobId)
      }
      state = .analysing(step: .calculating)

      let appError: AppError? = result.status == .failed
        ? .backendUnavailable(operation: "scoring")
        : nil
      if let appError {
        telemetry.error(appError, context: ["screen": "practice"])
        state = .error(appError)
        return
      }

      recordAttempt(jobId: job.jobId, score: result.score, durationMs: durationMs, status: .completed)
      state = .done(score: result.score, jobId: job.jobId)
      route.append(.result(
        jobId: job.jobId,
        surahId: surahId,
        ayahNumber: currentAyahNumber,
        score: result.score
      ))
    } catch let error as AppError {
      telemetry.error(error, context: ["screen": "practice"])
      recordAttempt(jobId: "", score: nil, durationMs: durationMs, status: .failed)
      state = .error(error)
      telemetry.event(
        TelemetryEvent.practiceScoringFailed,
        attributes: ["error_code": error.telemetryCode]
      )
    } catch {
      let appError = AppError.network(underlying: error)
      telemetry.error(appError, context: ["screen": "practice"])
      state = .error(appError)
    }
  }

  /// Reset back to idle so the user can retry. Drives the
  /// "Record again" CTA on the error panel.
  func resetForRetry() {
    state = .idle
  }

  /// Re-trigger the last reference playback as the "Replay" CTA shortcut.
  func replayLastRecording() {
    if case .ready = teacherState {
      toggleReferencePlayback()
    }
  }

  private func recordAttempt(
    jobId: String,
    score: Double?,
    durationMs: Int,
    status: AttemptStatus
  ) {
    let now = Date()
    historyStore.recordAttempt(Attempt(
      id: UUID().uuidString,
      surahId: surahId,
      surahNameEn: surah?.nameEn ?? surahId,
      ayahNumber: currentAyahNumber,
      score: score,
      jobId: jobId,
      createdAt: now,
      status: status,
      durationMs: durationMs
    ))
    if status == .completed, let score {
      historyStore.recordBestScore(
        surahId: surahId,
        ayahNumber: currentAyahNumber,
        score: score,
        achievedAt: now
      )
    }
    historyStore.setLastPracticed(LastPracticed(
      surahId: surahId,
      ayahNumber: currentAyahNumber,
      surahNameEn: surah?.nameEn ?? surahId,
      surahNameAr: surah?.nameAr ?? "",
      ayahCount: surah?.ayahCount ?? 0,
      practicedAt: now
    ))
  }

  private func observeRecorder() {
    // Mirror AudioRecorder's @Published meters/duration into the
    // PracticeRecordingState.recording associated values so the View
    // binds against the state machine directly.
    Task { @MainActor [weak self] in
      guard let self else { return }
      while case .recording = self.state {
        if self.recorder.isRecording {
          self.state = .recording(
            meters: self.recorder.meters,
            duration: self.recorder.duration
          )
        }
        try? await Task.sleep(nanoseconds: 50_000_000)
      }
    }
  }
}

File: apps/android/src/main/java/com/quran/android/features/practice/PracticeViewModel.kt
package com.quran.android.features.practice

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import com.quran.android.app.AppError
import com.quran.android.audio.Player
import com.quran.android.audio.Recorder
import com.quran.android.audio.RecordingResult
import com.quran.android.backend.AyahDetail
import com.quran.android.backend.QuranBackend
import com.quran.android.backend.ScoringResultPayload
import com.quran.android.backend.ScoringStatus
import com.quran.android.backend.SurahSummary
import com.quran.android.storage.Attempt
import com.quran.android.storage.AttemptStatus
import com.quran.android.storage.HistoryStore
import com.quran.android.storage.LastPracticed
import com.quran.android.telemetry.Telemetry
import com.quran.android.telemetry.TelemetryAttribute
import com.quran.android.telemetry.TelemetryEvent
import java.time.Instant
import java.util.UUID

/**
 * Practice tab state container. Mirrors
 * `apps/ios/.../Features/Practice/PracticeViewModel.swift`. Drives
 * the full Recording -> Uploading -> Analysing -> Done arc; PR 16
 * adds the analysing checklist UI + error panel; PR 17 wires the
 * Result navigation.
 */
class PracticeViewModel(
    private val backend: QuranBackend,
    private val recorder: Recorder,
    val player: Player,
    private val historyStore: HistoryStore,
    private val telemetry: Telemetry,
    private val surahId: String,
    private val ayahNumber: Int,
    private val pollAttempts: Int = DEFAULT_POLL_ATTEMPTS,
    private val pollIntervalMs: Long = DEFAULT_POLL_INTERVAL_MS,
) : ViewModel() {

    private val _state = MutableStateFlow<PracticeState>(PracticeState.Idle)
    val state: StateFlow<PracticeState> = _state.asStateFlow()

    private val _ayah = MutableStateFlow<AyahDetail?>(null)
    val ayah: StateFlow<AyahDetail?> = _ayah.asStateFlow()

    private val _surah = MutableStateFlow<SurahSummary?>(null)
    val surah: StateFlow<SurahSummary?> = _surah.asStateFlow()

    private val _reference = MutableStateFlow<TeacherReferenceState>(TeacherReferenceState.Loading)
    val reference: StateFlow<TeacherReferenceState> = _reference.asStateFlow()

    private var referencePlayedOnce: Boolean = false
    private var recorderObserverJob: Job? = null
    var lastRecording: RecordingResult? = null
        private set

    init {
        viewModelScope.launch { loadAyah() }
        viewModelScope.launch { loadSurah() }
        viewModelScope.launch { loadReference() }
    }

    /** Suspending helper exposed for unit tests. */
    suspend fun loadAyah() {
        try {
            _ayah.value = backend.ayah(surahId, ayahNumber)
        } catch (cause: AppError) {
            telemetry.error(cause, mapOf("screen" to "practice"))
            _state.value = PracticeState.Error(cause)
        }
    }

    suspend fun loadSurah() {
        try {
            _surah.value = backend.surah(surahId)
        } catch (cause: AppError) {
            telemetry.error(cause, mapOf("screen" to "practice", "section" to "surah"))
        }
    }

    suspend fun loadReference() {
        _reference.value = TeacherReferenceState.Loading
        try {
            val audio = backend.referenceAudio(surahId, ayahNumber)
            player.load(audio.signedUrl)
            _reference.value = TeacherReferenceState.Ready(player.state.value)
        } catch (cause: AppError.ReferenceUnavailable) {
            _reference.value = TeacherReferenceState.Unavailable(cause)
        } catch (cause: AppError) {
            telemetry.error(cause, mapOf("screen" to "practice", "section" to "reference"))
            _reference.value = TeacherReferenceState.Unavailable(
                AppError.ReferenceUnavailable(surahId, ayahNumber),
            )
        }
    }

    fun playReference() {
        player.play()
        if (!referencePlayedOnce) {
            referencePlayedOnce = true
            telemetry.event(
                TelemetryEvent.PRACTICE_REFERENCE_PLAYED,
                mapOf(
                    TelemetryAttribute.SURAH_ID to surahId,
                    TelemetryAttribute.AYAH to ayahNumber.toString(),
                ),
            )
        }
        refreshReferenceFromPlayer()
    }

    fun pauseReference() {
        player.pause()
        refreshReferenceFromPlayer()
    }

    fun setReferenceRate(rate: Float) {
        player.setRate(rate)
        refreshReferenceFromPlayer()
    }

    fun retryReference() {
        viewModelScope.launch { loadReference() }
    }

    fun startRecording() {
        viewModelScope.launch {
            try {
                player.pause()
                recorder.start()
                _state.value = PracticeState.Recording(meters = emptyList(), durationMs = 0L)
                telemetry.event(
                    TelemetryEvent.PRACTICE_RECORDING_STARTED,
                    mapOf(
                        TelemetryAttribute.SURAH_ID to surahId,
                        TelemetryAttribute.AYAH to ayahNumber.toString(),
                    ),
                )
                observeRecorderState()
            } catch (cause: AppError) {
                telemetry.error(cause, mapOf("screen" to "practice", "section" to "recording"))
                _state.value = PracticeState.Error(cause)
            }
        }
    }

    fun resetIdle() {
        recorder.cancel()
        _state.value = PracticeState.Idle
    }

    /** Convenience used by PracticeScreen's Replay action. */
    fun viewModelScopeLaunchUploadAndScore(recording: RecordingResult) {
        viewModelScope.launch { uploadAndScore(recording) }
    }

    fun stopRecording() {
        viewModelScope.launch {
            recorderObserverJob?.cancel()
            recorderObserverJob = null
            try {
                val result = recorder.stop()
                lastRecording = result
                telemetry.event(
                    TelemetryEvent.PRACTICE_RECORDING_STOPPED,
                    mapOf(TelemetryAttribute.DURATION_MS to result.durationMs.toString()),
                )
                uploadAndScore(result)
            } catch (cause: AppError) {
                telemetry.error(cause, mapOf("screen" to "practice", "section" to "recording"))
                _state.value = PracticeState.Error(cause)
            }
        }
    }

    /** Suspending helper exposed for unit tests; safe to call directly. */
    suspend fun uploadAndScore(recording: RecordingResult) {
        _state.value = PracticeState.Uploading
        try {
            val signed = telemetry.measure(TelemetryEvent.PRACTICE_UPLOAD_COMPLETED) {
                val payload = backend.requestSignedUploadUrl(
                    filename = recording.file.name,
                    contentType = AUDIO_M4A,
                )
                backend.uploadAudio(recording.file, payload.url, AUDIO_M4A)
                payload
            }
            val job = backend.createScoringJob(
                uploadKey = signed.uploadKey,
                surahId = surahId,
                ayahNumber = ayahNumber,
                sessionId = null,
            )
            val finalJob = telemetry.measure(TelemetryEvent.PRACTICE_SCORING_COMPLETED) {
                pollUntilTerminal(job)
            }
            recordHistory(finalJob, recording)
            _state.value = PracticeState.Done(score = finalJob.score, jobId = finalJob.jobId)
        } catch (cause: AppError) {
            telemetry.error(cause, mapOf("screen" to "practice", "section" to "scoring"))
            telemetry.event(
                TelemetryEvent.PRACTICE_SCORING_FAILED,
                mapOf(TelemetryAttribute.ERROR_CODE to cause.telemetryCode),
            )
            _state.value = PracticeState.Error(cause)
        }
    }

    private suspend fun pollUntilTerminal(initial: ScoringResultPayload): ScoringResultPayload {
        if (initial.status == ScoringStatus.Completed || initial.status == ScoringStatus.Failed) {
            return initial
        }
        for (attempt in 1..pollAttempts) {
            _state.value = PracticeState.Analysing(stepFor(attempt))
            delay(pollIntervalMs)
            val refreshed = backend.fetchScoringJob(initial.jobId)
            if (refreshed.status == ScoringStatus.Completed || refreshed.status == ScoringStatus.Failed) {
                return refreshed
            }
        }
        throw AppError.ScoringTimeout
    }

    private fun stepFor(attempt: Int): AnalysingStep = when {
        attempt <= 2 -> AnalysingStep.Transcribing
        attempt <= 5 -> AnalysingStep.Comparing
        else -> AnalysingStep.Calculating
    }

    private suspend fun recordHistory(
        job: ScoringResultPayload,
        recording: RecordingResult,
    ) {
        val now = Instant.now()
        val surahName = _surah.value?.nameEn ?: surahId
        val surahNameAr = _surah.value?.nameAr ?: ""
        val ayahCount = _surah.value?.ayahCount ?: 0
        val attempt = Attempt(
            id = UUID.randomUUID().toString(),
            surahId = surahId,
            surahNameEn = surahName,
            ayahNumber = ayahNumber,
            score = job.score,
            jobId = job.jobId,
            createdAt = now,
            status = if (job.status == ScoringStatus.Completed) AttemptStatus.COMPLETED else AttemptStatus.FAILED,
            durationMs = recording.durationMs,
        )
        historyStore.recordAttempt(attempt)
        historyStore.setLastPracticed(
            LastPracticed(
                surahId = surahId,
                ayahNumber = ayahNumber,
                surahNameEn = surahName,
                surahNameAr = surahNameAr,
                ayahCount = ayahCount,
                practicedAt = now,
            ),
        )
        job.score?.takeIf { job.status == ScoringStatus.Completed }?.let { score ->
            historyStore.recordBestScore(surahId, ayahNumber, score, now)
        }
    }

    private fun observeRecorderState() {
        recorderObserverJob?.cancel()
        recorderObserverJob = viewModelScope.launch {
            recorder.state.collect { rs ->
                if (_state.value is PracticeState.Recording) {
                    _state.value = PracticeState.Recording(rs.meters, rs.durationMs)
                }
            }
        }
    }

    private fun refreshReferenceFromPlayer() {
        if (_reference.value is TeacherReferenceState.Ready) {
            _reference.value = TeacherReferenceState.Ready(player.state.value)
        }
    }

    override fun onCleared() {
        player.release()
        super.onCleared()
    }

    companion object {
        const val DEFAULT_POLL_ATTEMPTS = 20
        const val DEFAULT_POLL_INTERVAL_MS = 2_000L
        const val AUDIO_M4A = "audio/m4a"
    }
}
